hcl/fmt.go

```diff
@@ -151,3 +164,318 @@ func (f FormatResult) Formatted() string {
+// the list the returned position will be equal to len(tokens).
+func fmtListExpr(tokens hclwrite.Tokens) (hclwrite.Tokens, int) {
+	logger := log.With().
+		Str("action", "hcl.fmtListExpr()").
+		Str("tokens", tokensStr(tokens)).
+		Logger()
+
+	logger.Trace().Msg("formatting list")
+
+	elemNextPos := 0
+	newTokens := hclwrite.Tokens{tokens[elemNextPos], newlineToken()}
+	elemNextPos++
+
+	skipNls := func() {
+		_, skipped := skipNewlines(tokens[elemNextPos:])
+		elemNextPos += skipped
+	}
+
+	for {
+		skipNls()
+
+		nextTokenType := tokens[elemNextPos].Type
+
+		if nextTokenType == hclsyntax.TokenComma {
+			elemNextPos++
+			continue
+		}
+
+		if nextTokenType == hclsyntax.TokenComment {
+			newTokens = append(newTokens, tokens[elemNextPos])
+			elemNextPos++
+			continue
+		}
+
+		if nextTokenType == hclsyntax.TokenCBrack {
+			logger.Trace().Msg("reached end of list")
+			break
+		}
+
+		logger.Trace().Msg("getting next element of the list")
+
+		element, nextPos := fmtNextElement(tokens[elemNextPos:])
+		elemNextPos += nextPos
+
+		logger.Trace().
+			Str("element", tokensStr(element)).
+			Str("tokens", tokensStr(tokens)).
+			Int("elemNextPos", elemNextPos).
+			Msg("new element got")
+
+		newTokens = append(newTokens, element...)
+		// Heredocs need to be handled differently, the comma must
+		// be on the next line in this case
+		if isHeredoc(element) {
+			newTokens = append(newTokens, newlineToken())
+		}
+		newTokens = append(newTokens, commaToken(), newlineToken())
+	}
+
+	newTokens = append(newTokens, closeBracketToken())
+	elemNextPos++
+
+	// Handling ["one"][0] and things like [[0]%[0]]
+	// We can also have newlines when dealing with operations
+
+	logger.Trace().Msg("checking if formatted list has operators/index access")
+
+	skipNls()
+
+	if elemNextPos == len(tokens) {
+		logger.Trace().Msg("no more tokens, returning formatted list")
+		return newTokens, elemNextPos
+	}
+
+	// We are handling things like this:
+	// var = [[] # c1\n\n #c\n [*]]
+	// We need to keep any comments after the immediate end of the list
+	// We don't keep the extra newlines, only newlines belonging to comments themselves.
+	hasCommentOrNl := true
+	for hasCommentOrNl {
+		switch tokens[elemNextPos].Type {
+		case hclsyntax.TokenComment:
+			logger.Trace().Msg("found comment after end of list, adding token")
+			newTokens = append(newTokens, tokens[elemNextPos])
+			elemNextPos++
+		case hclsyntax.TokenNewline:
+			logger.Trace().Msg("found newline after end of list, ignoring")
+			elemNextPos++
+		default:
+			hasCommentOrNl = false
+		}
+	}
+
+	nextTokenType := tokens[elemNextPos].Type
+
+	switch nextTokenType {
+	case hclsyntax.TokenComma, hclsyntax.TokenCBrack:
+		{
+			logger.Trace().Msg("end of outer list element, returning formatted list")
+			return newTokens, elemNextPos
+		}
+	case hclsyntax.TokenOBrack:
+		{
+			logger.Trace().Msg("getting tokens for list index access")
+
+			indexAccess, nextPos := fmtAnyExpr(tokens[elemNextPos:])
```

Comment:
since we already parsed/validated what would be the downside of just calling the function that handle a comma but the comma won't be there ? I explicitely didn't revalidated anything since I'm not doing parsing, in the sense of checking if the tokens make sense, I only check the bare minimum to introduce the newlines into the lists. If we go in the direction of validating things we would need to add a lot more checks and I don't think this make sense since we already called the hcl parser to check if the code was correct (it would make more sense then to just do our own parser etc and drop their parser entirely, if it is not reliable etc, but not what we are aiming here I think).